olp-cpp-sdk-core/src/cache/DefaultCache.cpp

```diff
@@ -195,58 +212,41 @@ boost::any DefaultCache::Get(const std::string& key, const Decoder& decoder) {
     }
   }
 
-  if (mutable_cache_) {
-    auto expiry = GetRemainingExpiryTime(key, *mutable_cache_);
+  bool valid = false;
+  DataPtr result;
+  time_t expiry = 0;
+
+  if (protected_cache_) {
+    std::tie(valid, result, expiry) = GetFromCache(key, *protected_cache_);
+  }
+
+  if (mutable_cache_ && !valid) {
+    std::tie(valid, result, expiry) = GetFromCache(key, *mutable_cache_);
+
     if (expiry <= 0) {
       PurgeDiskItem(key, *mutable_cache_);
-      return boost::any();
     }
+  }
 
-    auto value = mutable_cache_->Get(key);
-    if (value) {
-      auto decoded_item = decoder(value.get());
-      if (memory_cache_) {
-        memory_cache_->Put(key, decoded_item, expiry, value->size());
-      }
-      return decoded_item;
+  if (valid) {
+    auto decoded_item = decoder({result->begin(), result->end()});
+    if (memory_cache_) {
+      memory_cache_->Put(key, decoded_item, expiry, result->size());
     }
+    return decoded_item;
   }
 
   return boost::any();
 }
 
-std::shared_ptr<std::vector<unsigned char>> DefaultCache::Get(
+DataPtr DefaultCache::Get(
     const std::string& key) {
-  std::lock_guard<std::mutex> lk(cache_lock_);
-  if (!is_open_) {
-    return nullptr;
-  }
-
-  if (memory_cache_) {
-    auto value = memory_cache_->Get(key);
-
-    if (!value.empty()) {
-      return boost::any_cast<std::shared_ptr<std::vector<unsigned char>>>(
-          value);
-    }
-  }
-
-  if (mutable_cache_) {
-    auto expiry = GetRemainingExpiryTime(key, *mutable_cache_);
-    if (expiry <= 0) {
-      PurgeDiskItem(key, *mutable_cache_);
-      return nullptr;
-    }
+  auto data =
```

Comment:
Data copy is not good, just to avoid duplication! Please find a more effective way!

Reply:
Redone